### PR TITLE
feat(core): add GNAP persistent state layer for parallel agent task handoffs

### LINKED ISSUE #472: GNAP: persistent state layer for parallel coding agent task handoffs
Really solid parallel coding agent orchestrator. The task planning → spawning → parallel execution pattern is exactly right.

**[GNAP](https://github.com/farol-team/gnap)** could serve as the persistent task state backend for your orchestration pipeline.

**Current flow (in-memory):**
Planner creates tasks → spawns agents → agents execute in parallel → results collected

**With GNAP (persistent):**
```
gnap/tasks/
├── task-001.json  {title: "Implement auth", agent: "claude-1", status: "in-progress"}
├── task-002.json  {title: "Write tests", agent: "claude-2", status: "pending", depends_on: "task-001"}
└── task-003.json  {title: "Update docs", agent: "claude-3", status: "pending"}
```

**Why this matters for coding agents specifically:**
- Long-running coding tasks (hours) need persistence across agent restarts
- Multiple agents working on the same codebase need clear ownership (who's working on auth vs tests)
- Git-based task state fits naturally alongside the code git repo

**Minimal integration:**
The orchestrator writes GNAP tasks when spawning agents, agents update status on completion. Zero extra infrastructure — GNAP lives in the same git repo as the code.

Repo: https://github.com/farol-team/gnap

### PR DESCRIPTION
**Pull request by ashish921998**
## Summary

- Implements [GNAP (Git-Native Agent Protocol)](https://github.com/farol-team/gnap) v4 as an opt-in persistent state backend for the orchestration pipeline
- When enabled per-project via `gnap.enabled` config, task state is written as JSON files to `.gnap/` in the project's git repository
- Provides persistence across agent restarts, clear task ownership for parallel agents, and git-based audit trail

### New files
- `packages/core/src/gnap.ts` — Core GNAP module implementing all four protocol entities (agents, tasks, runs, messages) with read/write operations, state mapping from AO session states to GNAP task states, and sync functions
- `packages/core/src/__tests__/gnap.test.ts` — 60 tests covering all GNAP operations

### Modified files
- `packages/core/src/types.ts` — Added `gnap` config field to `ProjectConfig`
- `packages/core/src/config.ts` — Added Zod schema for GNAP config validation
- `packages/core/src/session-manager.ts` — Syncs to GNAP on session spawn (best-effort, never blocks spawn)
- `packages/core/src/lifecycle-manager.ts` — Syncs status transitions to GNAP task state (best-effort, never blocks polling)
- `packages/core/src/index.ts` — Exports all GNAP types and functions

### Integration pattern
```yaml
# agent-orchestrator.yaml
projects:
  my-app:
    repo: org/repo
    path: ~/my-app
    gnap:
      enabled: true
```

This writes GNAP-compliant state to the project repo:
```
.gnap/
├── version          # "4"
├── agents.json      # {session-id: {name, type, status}}
├── tasks/
│   └── INT-123.json # {title, assigned_to, state, ...}
├── runs/
│   └── INT-123-int-1.json  # {task_id, agent_id, state, ...}
└── messages/
```

## Design decisions
- **Best-effort only** — GNAP sync failures never block session spawn or lifecycle polling
- **Opt-in** — Disabled by default, enabled per-project via config
- **No git operations** — Writes files only; committing/pushing is left to the application layer
- **Deterministic task IDs** — Uses issue ID when available, session ID as fallback

Closes #472

## Test plan
- [x] 60 new tests covering all GNAP operations (init, task CRUD, agent CRUD, run CRUD, messages, state mapping, session sync, decomposition sync)
- [x] All 531 existing core tests pass
- [x] TypeScript type-checking passes
- [x] ESLint passes (0 new errors)

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: packages/core/src/gnap.ts [hunk 1]
```diff
@@ -0,0 +1,658 @@
...
+export function generateGnapTaskId(sessionId: string, issueId?: string): string {
+  if (issueId) {
+    // Sanitize issue ID for use as filename: replace slashes, spaces, special chars
+    return issueId
+      .replace(/[^a-zA-Z0-9_-]/g, "-")
+      .replace(/-+/g, "-")
+      .replace(/^-|-$/g, "");
+  }
+  return sessionId;
+}
```
[the remark this answers is not in the document]
Comment (ashish921998) [reply packages/core/src/gnap.ts:560]: Fixed in c7b2592. `generateGnapTaskId` now checks if the sanitized result is empty and falls back to `sessionId` when it is. Added tests for edge cases like `"#"`, `"///"`, and `"@"`.

File: packages/core/src/gnap.ts [hunk 2]
```diff
@@ -0,0 +1,658 @@
...
+      id: parentTaskId,
+      title: opts.rootTaskDescription,
+      assigned_to: [],
+      state: "in_progress",
+      created_by: "ao-decomposer",
+      created_at: now,
+      tags: ["decomposed"],
+    },
+    gnapDir,
+  );
```
[the remark this answers is not in the document]
Comment (ashish921998) [reply packages/core/src/gnap.ts:721]: Fixed in c7b2592. `syncDecompositionToGnap` now accepts an optional `planId` parameter. When not provided, it auto-generates a unique ID using `plan-{timestamp}`. Added a test verifying multiple decompositions don't collide.

File: packages/core/src/gnap.ts [hunk 3]
```diff
@@ -0,0 +1,665 @@
...
+
+/**
+ * Write a GNAP task file. Creates or overwrites the task file.
+ */
+export function writeGnapTask(projectPath: string, task: GnapTask, gnapDir = ".gnap"): void {
+  const root = getGnapDir(projectPath, gnapDir);
+  const tasksDir = getTasksDir(root);
+  mkdirSync(tasksDir, { recursive: true });
+
+  const taskPath = join(tasksDir, `${task.id}.json`);
```
[the remark this answers is not in the document]
Comment (ashish921998) [reply packages/core/src/gnap.ts]: Fixed in 5fe67ad. Added `validateGnapId()` and `safeFilePath()` that reject IDs containing path traversal sequences (`..`, `/`, `\\``) or empty strings. All write/read functions for tasks, runs, and messages now go through this validation. Also added `resolve()`-based containment check as a defense-in-depth measure. Added 3 tests for path traversal rejection.

**Comment by harshitsinghbhandari** [comment]
Closing this as part of the migration cleanup for the ground-up rewrite of
Agent Orchestrator.

This item was opened more than a month before the migration batch and the
codebase has been replaced, so this thread no longer maps cleanly onto the
current source tree.

If something here still needs tracking, please reopen this item or open a fresh
one against the new tree. Thank you for contributing.
